### PR TITLE
Fix tests when run without TTY

### PR DESCRIPTION
**Pull request by addaleax**
- Fix how the process/io mocking in `test/helpers/utils.js` handles synchronous uncaught exceptions by adding a `finally` handler resetting the mocked state.
- Skip two tests that would throw an exception when the test output is redirected to a non-TTY (e.g. `npm test | cat`).

Ref: https://github.com/nodejs/node/issues/8138

### CODE REVIEW

File: test/helpers/utils.js [hunk 1]
```diff
@@ -40,18 +40,27 @@ exports.checkOutput = function (f, argv, cb) {
 
     f()
   } else {
-    result = f()
+    try {
+      result = f()
+    } finally {
+      reset()
```
Comment (nexdrew) [inline test/helpers/utils.js]: If `f()` doesn't throw, `reset()` is called twice. But I guess twice is better than never. ;)

**Comment by nexdrew** [comment]
LGTM 👍 Thanks, @addaleax!

File: test/helpers/utils.js (hunk 1 above)
Comment (addaleax) [inline test/helpers/utils.js]: Yeah, figured as much… do you feel like this is something you’d like to see addressed or are you fine with it this way?

**Comment by addaleax** [comment]
Is the Windows failure something known or should I take a closer look at it?

**Comment by addaleax** [comment]
Oh… the Windows test failure is real, it just didn’t show up in CI because the Windows tests have the exact same problem CITGM was having, too.

File: test/helpers/utils.js (hunk 1 above)
Comment (nexdrew) [inline test/helpers/utils.js]: Seeing as how this code doesn't change much and the operation is idempotent, I'm fine leaving it this way.

**Comment by addaleax** [comment]
Ok, I’m out of luck for the current AppVeyor failure, that probably has to be left for someone with proper access to a Windows machine…

**Comment by maxrimue** [comment]
@addaleax I can look into it tomorrow :+1:

**Comment by addaleax** [comment]
Nice, thanks. :)

**Comment by maxrimue** [comment]
@addaleax I haven't fully figured out the problem yet, but I know for sure already that tests only fail in the CI, those two failing tests do not fail on my local machine, and they also fail on every Node.js from 0.10 to 6

**Comment by addaleax** [comment]
Turns out also skipping a third test for the combination of Windows & Node v0.10 does the trick. I think that’s okay? At least everything’s green now except for the known problems with Node 6.4.0! :tada:

**Comment by maxrimue** [comment]
@addaleax Great! :+1:

**Comment by nexdrew** [comment]
CI is green, I vote squash and merge

**Comment by addaleax** [comment]
Cool, thanks!

**Pull request merged**